src/optimagic/logging/logger.py

```diff
@@ -0,0 +1,320 @@
+        ],
+    ):
+        self.step_store = step_store
+        self.iteration_store = iteration_store
+        self.problem_store = problem_store
+        self._pytree_registry = get_registry(extended=True)
+
+    @cached_property
+    def _treedef(self) -> tuple[list[Any], Any]:
+        return tree_flatten(self.read_start_params(), registry=self._pytree_registry)[
+            -1
+        ]
+
+    def read_iteration(self, iteration: int) -> CriterionEvaluationWithId:
+        """Read a specific iteration from the iteration store.
+
+        Args:
+            iteration: The iteration number to read. Negative values read from the end.
+
+        Returns:
+            A `CriterionEvaluationWithId` object containing the iteration data.
+
+        Raises:
+            IndexError: If the iteration is invalid or the store is empty.
+
+        """
+        if iteration >= 0:
+            rowid = iteration + 1
+        else:
+            try:
+                last_row = self.iteration_store.select_last_rows(1)
+                highest_rowid = last_row[0].rowid
+            except IndexError as e:
+                raise IndexError(
+                    "Invalid iteration request, iteration store is empty"
+                ) from e
+
+            # iteration is negative here!
+            assert highest_rowid is not None
+            rowid = highest_rowid + iteration + 1
+
+        row_list = self.iteration_store.select(rowid)
+
+        if len(row_list) == 0:
+            raise IndexError(f"Invalid iteration requested: {iteration}")
+        else:
+            data = row_list[0]
+
+        params = tree_unflatten(
+            self._treedef, data.params, registry=self._pytree_registry
+        )
+
+        return replace(data, params=params)
+
+    def read_history(self) -> IterationHistory:
+        """Read the entire iteration history from the iteration store.
+
+        Returns:
+            An `IterationHistory` object containing the parameters,
+                criterion values, and runtimes.
+
+        """
+        raw_res = self.iteration_store.select()
+        params_list = []
+        criterion_list = []
+        runtime_list = []
+        for data in raw_res:
+            if data.value is not None:
+                params = tree_unflatten(
+                    self._treedef, data.params, registry=self._pytree_registry
+                )
+                params_list.append(params)
+                criterion_list.append(data.value)
+                runtime_list.append(data.timestamp)
+
+        times = np.array(runtime_list)
+        times -= times[0]
+
+        return IterationHistory(params_list, criterion_list, times)
+
+    @staticmethod
+    def _normalize_direction(
+        direction: OptimizationType | OptimizationTypeLiteral,
+    ) -> OptimizationType:
+        if isinstance(direction, str):
+            direction = OptimizationType(direction)
+        return direction
+
+    def _build_history_dataframe(self) -> pd.DataFrame:
```

Comment:
What are the usages of the history DataFrame? I wonder if we could get rid of that format as a DataFrame is not really a good way to store `params`, which can be anything.